plugins/eos/gs-plugin-eos.c

```diff
@@ -1553,3 +1873,191 @@ gs_plugin_add_popular (GsPlugin *plugin,
+	 * updates */
+	if (priv->os_upgrade && should_add_os_upgrade (priv->os_upgrade))
+		return TRUE;
+
+	/* poll so we eventually get a state change for the OS update, but never
+	 * return an error since we deal with them in the state change callback */
+	if (!eos_updater_call_poll_sync (priv->updater_proxy,
+					 cancellable, &local_error))
+		g_warning ("Error polling OS upgrade: %s", local_error->message);
+
+	return TRUE;
+}
+
+gboolean
+gs_plugin_add_distro_upgrades (GsPlugin *plugin,
+			       GsAppList *list,
+			       GCancellable *cancellable,
+			       GError **error)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	GsApp *os_upgrade = get_os_upgrade (plugin);
+
+	/* if we are testing the plugin, then always add the OS upgrade */
+	if (g_getenv ("GS_PLUGIN_EOS_TEST") != NULL) {
+		if  (gs_app_get_state (os_upgrade) == AS_APP_STATE_UNKNOWN)
+			gs_app_set_state (os_upgrade, AS_APP_STATE_AVAILABLE);
+		gs_app_list_add (list, os_upgrade);
+		return TRUE;
+	}
+
+	if (sync_state_from_updater (plugin) == EOS_UPDATER_STATE_READY) {
+		eos_updater_call_poll (priv->updater_proxy, cancellable, NULL,
+				       NULL);
+		return TRUE;
+	}
+
+	if (should_add_os_upgrade (os_upgrade)) {
+		g_debug ("Adding EOS upgrade: %s",
+			 gs_app_get_unique_id (os_upgrade));
+		gs_app_list_add (list, os_upgrade);
+	}
+
+	return TRUE;
+}
+
+gboolean
+gs_plugin_app_upgrade_download (GsPlugin *plugin,
+				GsApp *app,
+			        GCancellable *cancellable,
+				GError **error)
+{
+	g_autoptr(GError) local_error = NULL;
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	GsApp *os_upgrade = NULL;
+	gboolean needs_polling = FALSE;
+	EosUpdaterState eos_updater_state;
+
+	/* only process this app if was created by this plugin */
+	if (g_strcmp0 (gs_app_get_management_plugin (app),
+		       gs_plugin_get_name (plugin)) != 0)
+		return TRUE;
+
+	os_upgrade = get_os_upgrade (plugin);
+	if (app != os_upgrade) {
+		g_warning ("The OS upgrade to download (%s) differs from the "
+			   "one in the EOS plugin, yet it's managed by it!",
+			   gs_app_get_unique_id (app));
+		return TRUE;
+	}
+
+	eos_updater_state = sync_state_from_updater (plugin);
+	/* we need to poll again when there has been an error or it's one of the
+	 * initial states */
+	switch (eos_updater_state) {
+	case EOS_UPDATER_STATE_ERROR:
+	case EOS_UPDATER_STATE_READY:
+	case EOS_UPDATER_STATE_NONE:
+		needs_polling = TRUE;
+	default:
+		break;
+	}
+
+	os_upgrade_set_download_by_user (app, TRUE);
+
+	/* if it's already updatable (needs reboot) then there's nothing to be
+	 * done here */
+	if (gs_app_get_state (app) == AS_APP_STATE_UPDATABLE)
+		return TRUE;
+
+	gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+
+	/* poll or fetch the update, and the state of the OS upgrade will be
+	 * dealt with from outside this function, according to the state changes
+	 * of the update itself */
+	if (needs_polling) {
+		if (!eos_updater_call_poll_sync (priv->updater_proxy, cancellable,
```

Comment:
Because the `polling` state doesn't change the state, so we need to start the change here since we know that the user is explicitly downloading it.
I added the `download-by-user` metadata after most of this logic so I didn't change what I had. But I can use this metadata in the `sync_state_from_updater` if this is important.